{% extends 'report.html' %}
{% block report %}
<div>
  <h4>Download</h4>
  <ul>
    {% if offset_path %}<li><a href="{% url 'download' offset_path %}" download>A-site_offsets.tab</a></li>{% endif %}
    {% if perc_gene_path %}<li><a href="{% url 'download' perc_gene_path %}" download>Perc_of_genes_for_all_offsets.tab</a></li>{% endif %}
    {% if profile_path %}<li><a href="{% url 'download' profile_path %}" download>A-site_profiles.tab</a></li>{% endif %}
    {% if log_path %}<li><a href="{% url 'download' log_path %}" download>aip.log</a></li>{% endif %}
  </ul> 
</div>
{% if job.status == "SUCCESS" %}
<h4>Table 1. Most probable Offsets for Fragment Size and Frame</h4>
<table class="table table-bordered" id="offsets">
  <thead>    
    <th>Fragment size</th>
    <th>Frame 0</th>
    <th>Frame 1</th>
    <th>Frame 2</th>
  </thead>
</table>
<h4>Figure 1. Distribution of offset value</h4>
<p class="text-muted">By default, all fragment sizes are shown in the following figures. You can exclude a fragment size by clicking on the legend for that fragment size. </p>
<div class="row">
  <div class="col-sm-4 offset-hist" style="height:350px;"></div>
  <div class="col-sm-4 offset-hist" style="height:350px;"></div>
  <div class="col-sm-4 offset-hist" style="height:350px;"></div>
</div>
<h4>Figure 2. Distribution of mRNA fragment size</h4>
<div id="frag-size-hist" style="width:600px;height:400px;"></div>
<h4>Table 2. Number of genes for various fragment size and frame combinations that meet the criteria of at least 1 read per codon on average</h4>
<table class="table table-bordered" id="gene-dist">
  <thead>    
    <th>Fragment size</th>
    <th>Frame 0</th>
    <th>Frame 1</th>
    <th>Frame 2</th>
  </thead>
</table>
{% endif %}
<div>
  <h4>Parameters</h4>
  <ul>
      <li>Species: {{species}}</li>
      <li>Alignment type: {{job.alignment_type}} </li>
      <li>Fragment length: min {{job.min_frag}} nt, max {{job.max_frag}} nt</li>
      <li>Quantify reads from: {% if job.three_prime %} 3 {% else %} 5 {% endif%}' end</li>
      <li>Overlap: {{job.overlap}} nt</li>
      <li>Minimum average number of reads per codon for filtering genes: {{job.threshold_avg_reads}}</li>
      <li>Minimum percentage of genes with the most probable offset for assigning a unique offset: {{job.threshold_gene_pct}}%</li>
      <li>Minimum ratio between the average reads in the second, third and fourth codon and the reads in the first codon: {{job.threshold_start_codon}}</li>
  </ul>
</div>
{% endblock %}

{% block javascript %}
<script>
    $('#nav-job').addClass("active");
</script>
{% if job.status == "SUCCESS" %}
<script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
<script>
$(function(){
    $.ajax({
        url: '/ajax/get_offset_results/' + {{job.id}},
        dataType: 'json',
        success: function (data) {
          if (true) {
              // distribution of mRNA fragment size
              layout = {
                  xaxis: {
                      title:{
                          text: 'Fragment Size (nucleotides)'                              
                      },
                      tickmode: 'linear',
                      dtick: 1
                  },
                  yaxis: {
                      title: {
                          text: '% of total reads'
                      }  
                  }
              }
              Plotly.newPlot($('#frag-size-hist')[0], data["frag_size_hist"], layout);
              
              // distribution of offset value
              for (i = 0; i < 3; i++) {
                  layout = {
                      title:{
                          text: 'Frame ' + i
                      },
                      xaxis: {
                          title:{
                              text: 'Offset (nucleotides)'                              
                          },
                          tickmode: 'linear',
                          tick0: 0,
                          dtick: 3
                      },
                      annotations: [{
                        x: 1.25,
                        y: 1.1,
                        xref:'paper',
                        yref:'paper',
                        text:'Fragment Size',
                        showarrow: false
                      }]
                  }
                  if (i == 0) {
                      layout['yaxis'] = {
                          title: {
                              text: 'Genes with offset (%)'
                          }
                      };
                  }
                  Plotly.newPlot($('.offset-hist')[i], data["offset_pct"][i], layout);
              }

              // table for offsets
              $('#offsets').DataTable( {
                  data: data['offsets'],
                  searching: false,
                  scrollY: '50vh',
                  scrollCollapse: true,
                  paging: false,
                  bInfo: false
              } );

              $('#offsets tr>td:not(:first-child').each(function(index) {
                  var txt = $(this).text();
                  if(txt != "NA") {
                      if (txt.includes("/")) {
                          $(this).css("background-color", "#fff7e6");
                      } else {
                          $(this).css("background-color", "#e6ffe6");
                      }
                  }
              });
              
              // table for gene distribution
              $('#gene-dist').DataTable( {
                  'data': data['genes'],
                  "searching": false,
                  scrollY: '50vh',
                  scrollCollapse: true,
                  paging: false,
                  bInfo: false
              } );
          }
        }
      });
});

</script>
{% endif %}
{% endblock %}
