{% extends 'base.html' %}
{% load file_list %}
{% block css %}
<style>
.error {
    color: red;
}
</style>
{% endblock %}
{% block content %}
<h2>Get A-site Profiles</h2>
<form action="{% url 'submit_profile' %}" method="post" enctype="multipart/form-data" class="form">
  {% csrf_token %}
  <div class="form-group">
    <label>Description</label>
    {{form.name}}
    <span class="error">{{form.name.errors}}</span>
  </div>
  <div class="form-group">
    <label>Species</label>
    {{form.species}}
    <span class="error">{{form.species.errors}}</span>
  </div>
  <div class="form-group">
    <label>Alignment mode</label>
    <div class="form-check">
      <input class="form-check-input" type="radio"  name="alignment_type" value="genome" checked> 
      <label class="form-check-label">Genome</label> 
      <span class="form-text">This option is preferable for spaner genomes (e.g. <i>E. coli</i>, <i>S. cerevisiae</i>) with simple gene structures and fewer introns and isoforms.</span>
    </div>
    <div class="form-check">
      <input class="form-check-input" type="radio" name="alignment_type" value="transcriptome"> 
      <label class="form-check-label">Transcriptome</label>
      <span class="form-text">This option is preferable for complex genomes like mouse and human. The alignment file contains reads directly aligned to gene transcripts.</span>
    </div>
    <span class="error">{{form.alignment_type.errors}}</span>  
  </div>
  <div class="form-group">
    <label>BAM or SAM file</label>
    <select name="bam_file">
      {% file_list user "BAM" %}
    </select>
    <span class="error">{{form.bam_file.errors}}</span>    
  </div>
  <div class="form-group">
    <label>Annotation file</label>
    <select name="annotation_file">
      {% file_list user "Other" %}
    </select>
    <span class="error">{{form.annotation_file.errors}}</span>
    <div class="form-text" id="genome-annotation">
      <p>For genome, the annotation file could be a .tab file with the following fields (header is not needed in the .tab file):</p>
      <table class="table table-bordered">
        <thead>
          <th>Gene name</th>
          <th>Chromosome</th>
          <th>Strand</th>
          <th>CDS Length</th>
          <th>Number of CDS regions</th>
          <th>CDS Start 1</th>
          <th>CDS End 1</th>
          <th>CDS Start 2</th>
          <th>CDS End 2</th>
        </thead>
        <tbody>
          <tr>
            <td>YAL008W</td>
            <td>chrI</td>
            <td>+</td>
            <td>597</td>
            <td>1</td>
            <td>136914</td>
            <td>137510</td>
            <td></td>
            <td></td>
          </tr>
          <tr>
            <td>YBL087C</td>
            <td>chrII</td>
            <td>-</td>
            <td>414</td>
            <td>2</td>
            <td>59822</td>
            <td>60193</td>
            <td>60698</td>
            <td>60739</td>
          </tr>
        </tbody>
      </table>
      <p>Alternatively, a GFF file can also be given as input for annotations, but the script is optimized for processing GFF files for <i>E. coli</i> and <i>S. cerevisiae</i> only.</p>
    </div> 
    <div class="form-text" id="trans-annotation" style="display: none">
      <p>For transcriptome, the annotation file could be a .tab file with the following fields (header is not needed in the .tab file):</p>
      <table class="table table-bordered">
        <thead>
          <th>Gene name</th>
          <th>Start index</th>
          <th>CDS Length</th>
        </thead>
        <tbody>
          <tr>
            <td>uc008jxs.1</td>
            <td>79</td>
            <td>1155</td>
          </tr>
          <tr>
            <td>uc009fyo.2</td>
            <td>1048</td>
            <td>4023</td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div class="form-group">
    <label>FASTA file</label>
    <select name="fasta_file">
      {% file_list user "FASTA" %}
    </select>
    <span class="error">{{form.fasta_file.errors}}</span>  
    <p class="form-text" id="fasta-genome">Select the fasta file containing the genome sequence.</p>
    <p class="form-text" id="fasta-trans" style="display: none">Select the fasta file containing the transcriptome sequence.</p>
  </div>
  <div class="form-group">
    <label>Fragment size</label>
    min <input type="number" name = "min_frag" value="20" style="width:4em"> nt, 
    max <input type="number" name="max_frag" value="35" style="width:4em"> nt
    <span class="error">{{form.min_frag.errors}}</span>  
    <span class="error">{{form.max_frag.errors}}</span>
    <p class="form-text">The size of the read obtained in the Ribo-Seq Next-Generation Sequencing (NGS) experiment that represents a copy of the mRNA fragment protected by the ribosome and cleaved by the RNase enzyme.</p>
  </div>
  <div class="form-group">
    <label>Quantify reads from </label>
    <span><input type="radio" name="three_prime" value="false" checked> 5' or <input type="radio" name="three_prime" value="true"> 3' end</span>
    <span class="error">{{form.three_prime.errors}}</span>  
  </div>
  <div class="form-group">
    <label class="control-label">Overlap</label>
    <input type="number" name="overlap" value="0" style="width:4em"> nt
    <span class="error">{{form.overlap.errors}}</span>  
    <p class="form-text"> Number of nucleotides beyond the CDS region of a gene which are to be avoided to overlap with another gene. </p>
  </div>
  <div class="form-group"> 
    <div id="offset-file">
      <label>Offset file</label>
      <select name="offset_file">
        {% file_list user "Other" %}
      </select>
      <span class="error">{{form.offset_file.errors}}</span>  
      <p class="form-text">The offset file should contain offsets in rows of different fragment sizes and columns in frame 0,1 and 2. The first column should be fragment size. </p>
    </div>
  </div>
  <div class="form-group">
    <label>Email <input type="email" name="email" value="{{user.email}}"></label>
    <p class="form-text">A notification will be sent to this email address once the job finishes.</p>
    <span class="error">{{form.email.errors}}</span>
    <p class="form-text"></p>
  </div>
  <div class="form-group"> 
    <button type="submit" class="btn btn-primary">Submit</button>
  </div>
</form>
{% endblock %}

{% block javascript %}
<script>
  $("#nav-profile").addClass("active");
    
  $('select[name=species]').change(function(){
    if (this.value == "yeast") {
      $('select[name=annotation_file]').val('/files/input/default/Other/sacCer3_CDS.tab');
      $('select[name=fasta_file]').val('/files/input/default/FASTA/sacCer3_R64-2-1_genome.fa');
    } else if (this.value == "ecoli") {
      $('select[name=annotation_file]').val('/files/input/default/Other/ecoli_CDS.tab');
      $('select[name=fasta_file]').val('/files/input/default/FASTA/ecoli_nc_000913.3.fa');
    } else if (this.value == "mouse") {
      $('select[name=annotation_file]').val('/files/input/default/Other/mm10_start_index.tab');
      $('select[name=fasta_file]').val('/files/input/default/FASTA/mm10_transcript.fa');    
    }
  });
    
  $('input[type=radio][name=alignment_type]').change(function(){
      if (this.value == "genome") {
          $('#genome-annotation').show();
          $('#trans-annotation').hide();
          $('#fasta-genome').show();
          $('#fasta-trans').hide();
      } else {
          $('#genome-annotation').hide();
          $('#trans-annotation').show();
          $('#fasta-genome').hide();
          $('#fasta-trans').show();
      }
  });

  $('input[type=radio][name=get_profile]').change(function(){
      if (this.value == "true") {
          $('#offset-file').show();
      } else {
          $('#offset-file').hide();
      }
  })
  
</script>
{% endblock %}